### PR TITLE
Filestore throttle

### PR DESCRIPTION
**Pull request by majianpeng**
a)make wbthrottle act on clone/clone_range/clone_range2
b)fix bugs about wbthrottle
c)fix bugs or cleanup for filejournal throttle.

### CODE REVIEW

**Comment by majianpeng** [comment]
@liewegas .  I have a question: why use throttle for filejournal?  IMO,  in prepare_multi_write, we already check journal_max_write_entries/journal_max_write_bytes. I think this is the same as throttle. Or am i missing something?

File: src/os/FileStore.cc [hunk 1]
```diff
@@ -1891,8 +1915,16 @@ void FileStore::_do_op(OpSequencer *osr, ThreadPool::TPHandle &handle)
     dout(5) << "_do_op done stalling" << dendl;
   }
 
-  osr->apply_lock.Lock();
-  Op *o = osr->peek_queue();
+  do {
+    osr->apply_lock.Lock();
+    o = osr->peek_queue();
+    if (o->flags & OP_SKIP_WBTHROTTLE == 0) {
+      osr->apply_lock.Unlock();
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: why unlock apply_lock here?

File: src/os/FileStore.cc [hunk 2]
```diff
@@ -1891,8 +1915,16 @@ void FileStore::_do_op(OpSequencer *osr, ThreadPool::TPHandle &handle)
...
+      wbthrottle.throttle();
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: Looks "OP_SKIP_WBTHROTTLE" reduce a extra throttle? Is this really necessary to change?

Comment (yuyuyu101) [inline src/os/FileStore.cc]: Because we add a extra logic to check this

File: src/os/FileStore.cc [hunk 3]
```diff
@@ -3579,7 +3615,8 @@ int FileStore::_clone_range(coll_t cid, const ghobject_t& oldoid, const ghobject
   if (r < 0) {
     r = -errno;
     goto out3;
-  }
+  } else if (!replaying && g_conf->filestore_wbthrottle_enable)
+      wbthrottle.queue_wb(n, newoid, dstoff, r, false);
```
Comment (yuyuyu101) [inline src/os/FileStore.cc]: indent

**Comment by yuyuyu101** [comment]
I don't think journal_max_write_entries plays the same role with FileJournal throttle, the previous one only avoid one potential aio op doesn't contain too much front io, the latter block client io in case of queueing too much in FileJournal causing hang

File: src/os/FileStore.cc (hunk 2 above)
Comment (majianpeng) [inline src/os/FileStore.cc]: @yuyuyu101 . The function of wbthrottle is to control page cache of kernel. So it should not control no-write ops. Further, we already throttle_ops & throttle_bytes. As for benefit, this depend on workload. IIRR, for example mds store file entry w/  omap operation.

**Comment by majianpeng** [comment]
@yuyuyu101. Why does make FileJournal hang if there are much entry?

File: src/os/FileStore.cc (hunk 2 above)
Comment (yuyuyu101) [inline src/os/FileStore.cc]: Why xattr and omap aren't part of page cache throttler?

File: src/os/BtrfsFileStoreBackend.cc [hunk 4]
```diff
@@ -540,7 +543,7 @@ int BtrfsFileStoreBackend::clone_range(int from, int to, uint64_t srcoff, uint64
   a.dest_offset = dstoffclone;
   err = ::ioctl(to, BTRFS_IOC_CLONE_RANGE, &a);
   if (err >= 0) {
-    r += err;
+    r += 0; //expect this.It don't do anycopy in kernel.
```
Comment (varadakari) [inline src/os/BtrfsFileStoreBackend.cc]: why do we need this? you can remove the line and add comment.

Comment (varadakari) [inline src/os/BtrfsFileStoreBackend.cc]: we can change the if, handle in better way for err > 0 scenarios.

Comment (majianpeng) [inline src/os/BtrfsFileStoreBackend.cc]: In fact, i modify the meaning of return value for this function. So i use 'r += 0' and a comment to note this.

File: src/os/FileStore.cc (hunk 2 above)
Comment (majianpeng) [inline src/os/FileStore.cc]: In fact, only FileStore::_write count into wbthrottle. For xattr & omap don't count this.
For xattr/omap write-operations, it don't count into wbthrottle.

File: src/os/FileStore.cc [hunk 5]
```diff
@@ -1796,11 +1796,33 @@ FileStore::Op *FileStore::build_op(list<Transaction*>& tls,
...
+    if (!need_wbthrottle) {
+      Transaction::iterator i = (*p)->begin();
+      while (i.have_op()) {
+	Transaction::Op *op = i.decode_op();
+	switch(op->op) {
+	  case Transaction::OP_WRITE:
+	  case Transaction::OP_CLONE:
+	  case Transaction::OP_CLONERANGE:
+	  case Transaction::OP_CLONERANGE2:
+#if !defined(CEPH_HVAE_FALLOCATE)
```
Comment (varadakari) [inline src/os/FileStore.cc]: typo? CEPH_HAVE_FALLOCATE?

File: src/os/BtrfsFileStoreBackend.cc (hunk 4 above)
Comment (varadakari) [inline src/os/BtrfsFileStoreBackend.cc]: May be we can add a better comment and refactor if. If we are hitting it every time, we can refactor to reduce one branching, else add __likely()

File: src/os/FileStore.cc [hunk 6]
```diff
@@ -1796,11 +1796,33 @@ FileStore::Op *FileStore::build_op(list<Transaction*>& tls,
...
+	  case Transaction::OP_ZERO:
+#elif defined(DARWIN) || defined(__FreeBSD__)
```
Comment (varadakari) [inline src/os/FileStore.cc]: Can we combine #if and #elif to one condition, checking for the same op.

**Comment by yuyuyu101** [comment]
Not sure it worth to change these

**Comment by liewegas** [comment]
Yeah, I don't think this is worth the churn for FileStore.  However, we should keep the Throttle changes, and that is used all over the place.  And, ideally, switch to std::mutex at the same time...

**Pull request closed without merging**